### ISSUE TITLE
Start time much longer after upgrade from 6.x to 7.x

### ISSUE THREAD

**Issue by peacand** [elastic/logstash#11105]
Hi,

I've upgraded my Logstash instances from 6.8 to 7.3. After the upgrade, I've noticed a startup time ~3x higher, going from ~3min to ~9min. I'm surprised because I red in the release notes that the startup time is expected to be much faster with the new Java engine.

To give you more details about what I observe, the startup behavior is actually different from 6 to 7.

In version 6, during startup, in the logs, I had nothing for ~3min and then in a few seconds, I had all these lines and Logstash was 100% ready and processing logs :

```
Starting UDP listener {:address=>"0.0.0.0:xxxxx"}
UDP listener started {:address=>"0.0.0.0:xxxxx", :receive_buffer_bytes=>"106496", :queue_size=>"2000"}
Starting UDP listener {:address=>"0.0.0.0:yyyyy"}
UDP listener started {:address=>"0.0.0.0:yyyyy", :receive_buffer_bytes=>"106496", :queue_size=>"2000"}
....
Pipelines running {:count=>1, :running_pipelines=>[:main], :non_running_pipelines=>[]}
Opening file {:path=>"/var/log/xxxxxx"}
....
```

But in version 7 it's different. In ~1min after the startup command, I get these logs :

```
Starting UDP listener {:address=>"0.0.0.0:xxxxx"}
UDP listener started {:address=>"0.0.0.0:xxxxx", :receive_buffer_bytes=>"106496", :queue_size=>"2000"}
Starting UDP listener {:address=>"0.0.0.0:yyyyy"}
UDP listener started {:address=>"0.0.0.0:yyyyy", :receive_buffer_bytes=>"106496", :queue_size=>"2000"}
....
Pipelines running {:count=>1, :running_pipelines=>[:main], :non_running_pipelines=>[]}
```

So at that time it looks like the pipeline is ready. But in fact not. There is no output at that time. The output starts coming ~8 minutes later. And during this time where the inputs are ready, the pipeline is supposed to run but there is no output. I've strace the process and it's actually generating all the Java CompileDataset in /tmp. In my case there are 400 files like that to generate. After the last file is generated, the output starts coming.

So it looks like the pipeline is ready in ~1min, but in fact it takes really 9min.

Do you know why it takes so long ? Did I miss a configuration option while upgrading to 7.x ?
In my pipeline I use many custom Ruby filters, both inline and using external Ruby script files.
Could it explain this high Java generation time ?

Thank you for your help

**Comment by sysmonk** [comment]
I'm basically having the same issue. I've upgraded from LS5 to LS7.3 and the startup time is terrible. The worst problem is, though, is that the input / pipeline is ready pretty fast, but the compile takes very long, and during that time a lot of things are stuck until the compilation of rules finishes. Our log shppers (log-courier) are connected (because pipeline starts accepting input)  and are trying to send  logs to the stuck Logstash servers, and there's more than 10k logs/s being sent, which gets LS server clogged up fast and stales the shipping to other Logstash servers.

Our worker count is *3 CPU count, and I've noticed that servers with LESS cpu's are finishing the compilation a lot faster than the servers with MORE CPU's.
So, on server with 8 CPU's, the worker count is 24, and the compilation finished in ~5 minutes.
On server with 24 CPU's, the worker count is 72, and the compilation finishes in ~14 minutes.
If i set the worker count to 24 on the 24 CPU's node, it also finishes in ~5 minutes. 

We have a bunch or rules, and there's around 494 CompiledDatasets files in the end.

Is there a possibility to pre-compile the rules to the java classes to speed up logstash restarts?

**Comment by sysmonk** [comment]
Got even more annoying, sometimes if Logstash crashes, we restart it, and get over 5000 servers connected too it and sending payloads, but as the server hasn't compiled the rules yet, the server  timesout, and stalls the 5000+ servers log sending, and also overloads the Logstash server, so rule compilation takes even longer :( 

Can we please get the pipeline started only AFTER the rules are compiled? Or at least have a setting for that? 

Thanks

**Comment by sysmonk** [comment]
@colinsurprenant  any luck on this or any workarounds? this is very very annoying :(

**Comment by colinsurprenant** [comment]
Thanks for the reports @peacand @sysmonk.

To help investigate the issue, the best would be to be able to replicate locally. Could share your configurations with us? Better yet, if we could come up with a simplified configuration which only uses stdin/stdout input and output that would also really help. 

Let me know if that is possible.

**Comment by colinsurprenant** [comment]
- Could you also provide information about your respective environment? OS, JVM, container?
- Did you try turning off the java pipeline to see if it changes anything? (using `pipeline.java_execution: false` in logstash.yml or using `--java-execution=false` on the command line. 
- There are a few related issues about the use of substitution variables in pipelines config; #10126 #10792 #10794 - would like your feedback to see if these likely apply to your problem.

**Comment by peacand** [comment]
Hi @colinsurprenant 

Thank you for your help. Unfortunately I cannot share my pipeline configuration, it contains sensitive information. 

But thanks to your suggestion I've understood the high latency is introduced by the new Java execution engine ! Using `pipeline.java_execution: false`, the start/restart time goes back as it was before the upgrade. It goes from ~11 minutes (with the Java engine) to ~4 minutes (without). 

Maybe I can describe my pipeline a little bit. It is composed of 47 configuration files, with a total of ~3000 lines. Most if it is actually conditions based on exact match and also many regex. The main actions are drop or add_field. We are also using much custom Ruby scripts. 

Based on that, do you have an idea why the Java execution engine is so slow to get the pipeline 100% ready ? Maybe all the regex which need to be compiled ? The Ruby scripts ? something else ... ?

Moreover, is there any risk turning off the Java engine ?

**Comment by colinsurprenant** [comment]
@peacand 

- Would it be possible for you to sanitize your configuration? I understand that this might represent some work but it would be useful. 
- Did you verify about my comment on the use of substitution variables in pipelines?
- There is no *risk* in turning off the java engine but other than the startup time, the java engine *should* be more efficient. The legacy Ruby engine will be eventually deprecated so the java engine is the way forward.

**Comment by sysmonk** [comment]
We don't use substituitions, so that should be realted. I've just restarted logstash with debug enabled and what i see is that logstash tries to compile everythiing at least $worker times. In my case - i've got  24 workers, so it compiles 24 times _at least_ :

`# grep 'org.logstash.config.ir.compiler.ComputeStepSyntaxElement' logstash-startup.txt | sort | uniq -c | sort -n | head -10
     24  org.logstash.config.ir.compiler.ComputeStepSyntaxElement@100469b2
     24  org.logstash.config.ir.compiler.ComputeStepSyntaxElement@10deede6
     24  org.logstash.config.ir.compiler.ComputeStepSyntaxElement@10f9afb0
     24  org.logstash.config.ir.compiler.ComputeStepSyntaxElement@111bcfa4
     24  org.logstash.config.ir.compiler.ComputeStepSyntaxElement@114e7908
     24  org.logstash.config.ir.compiler.ComputeStepSyntaxElement@1230b7c7
     24  org.logstash.config.ir.compiler.ComputeStepSyntaxElement@1246aefb
     24  org.logstash.config.ir.compiler.ComputeStepSyntaxElement@124c7c0d
     24  org.logstash.config.ir.compiler.ComputeStepSyntaxElement@12cf4a6e
     24  org.logstash.config.ir.compiler.ComputeStepSyntaxElement@137e7ae8`


But sometimes it compilels stuff unimanagablel times:

`grep 'org.logstash.config.ir.compiler.ComputeStepSyntaxElement' logstash-startup.txt | sort | uniq -c | sort -n | tail -10
     48  org.logstash.config.ir.compiler.ComputeStepSyntaxElement@5cf7e8bc
     48  org.logstash.config.ir.compiler.ComputeStepSyntaxElement@da57e3ac
     48  org.logstash.config.ir.compiler.ComputeStepSyntaxElement@fa2300ae
     72  org.logstash.config.ir.compiler.ComputeStepSyntaxElement@5e08af85
     72  org.logstash.config.ir.compiler.ComputeStepSyntaxElement@65be327e
     72  org.logstash.config.ir.compiler.ComputeStepSyntaxElement@9e84a9bd
    192  org.logstash.config.ir.compiler.ComputeStepSyntaxElement@54319efa
    216  org.logstash.config.ir.compiler.ComputeStepSyntaxElement@27ffd08
    360  org.logstash.config.ir.compiler.ComputeStepSyntaxElement@4b6b5538
    432  org.logstash.config.ir.compiler.ComputeStepSyntaxElement@b8bdbae1`


In this  case, b8bdbae1 is:
`grep -A10 -B10 b8bdbae1 logstash-startup.txt  | head -20
 into 
 org.logstash.config.ir.compiler.ComputeStepSyntaxElement@39dec1b4
[2019-12-19T17:05:42,666][DEBUG][org.logstash.config.ir.CompiledPipeline] Compiled conditional
 [if (event.getField('[sampleType]')=='COUNTERSSAMPLE')] 
 into 
 org.logstash.config.ir.compiler.ComputeStepSyntaxElement@f33fb33b
[2019-12-19T17:05:42,683][DEBUG][org.logstash.config.ir.CompiledPipeline] Compiled conditional
 [if (event.get('[srcIP6]'))] 
 into 
 org.logstash.config.ir.compiler.ComputeStepSyntaxElement@b8bdbae1
[2019-12-19T17:05:42,728][DEBUG][org.logstash.config.ir.CompiledPipeline] Compiled conditional
 [if (event.get('[srcIP6]'))] 
 into 
 org.logstash.config.ir.compiler.ComputeStepSyntaxElement@b8bdbae1
[2019-12-19T17:05:42,734][DEBUG][org.logstash.config.ir.CompiledPipeline] Compiled conditional
 [if (event.get('[srcIP6]'))] 
 into 
 org.logstash.config.ir.compiler.ComputeStepSyntaxElement@b8bdbae1
[2019-12-19T17:05:42,741][DEBUG][org.logstash.config.ir.CompiledPipeline] Compiled conditional`

This is the only two times the srcIP6 is mentioned in the configs:

`# grep srcIP6 *
20-sflow.conf:          else if [srcIP6] {
20-sflow.conf:                          rename => { "srcIP6" => "[netflow][ipv6_src_addr]" }`

The config basically is:
`filter {
        if "sflow" == [type] {

                json { }
                mutate { some fields into integers } 
   if [sampleType] == "COUNTERSSAMPLE" {
                        drop {}
                }
                else if [srcIP6] {
                        ruby {
                                code => 'event.set("[netflow][last_switched]", event.get("unixSecondsUTC") * 1000)'
                }
               mutate {
                                id => "sflow_mutate_sflow2netflow_ipv6_rename"
				rename some  fields...
	                                rename => { "srcIP6" => "[netflow][ipv6_src_addr]" }
				rename some fields
		}
                       if [ICMPType] {
                                ruby {
                                        code => 'event.set("[netflow][icmpTypeCodeIPv6]", (event.get("[ICMPType]") * 256) + event.get("[ICMPCode]"))'
                                }
                        }
                }
                else if [srcIP] {
                        ruby {
                                code => 'event.set("[netflow][last_switched]", event.get("unixSecondsUTC") * 1000)'
                        }
                        mutate {
			rename more
			}
                      if [ICMPType] {
                                ruby {
                                        code => 'event.set("[netflow][icmpTypeCodeIPv4]", (event.get("[ICMPType]") * 256) + event.get("[ICMPCode]"))'
                                }
                        }
                }
                else {
                        drop {}
                }
                ruby {
                        code => 'event.set("[netflow][in_bytes]", event.get("[IPSize]") * event.get("[netflow][in_pkts]"))'
                }
                mutate {
                        id => "sflow_mutate_sflow2netflow_remove_unused_fields"
                        remove_field => [ 
				remove_Some_Fields
			]
                mutate {
                        update => { "type" => "netflow" }
                        add_tag => ['sflow']
                }
        }`

I have no idea why it has to  compile the same rulles 24 time, i have even less idea why it sometimes needs to do that 432 times.
Btw, it's consistent over restarts.

**Comment by sysmonk** [comment]
thart's just part of the onfig, whidh is around 3600 lilnes ( multiple files)

**Comment by peacand** [comment]
@colinsurprenant 

To answer your questions :

- No I double checked and I really cant share the pipeline configuration. Sorry
- I don't use variables substitutions in the pipeline

I'll try to remove the custom Ruby scripts from the pipeline to see if it is the origin of the latency.

**Comment by colinsurprenant** [comment]
I have spent some time investigating this problem and was able to replicate quite easily using a very simple config with a large (50x) repetition of an `if` statement in the `filter {}` section.

The compilation of a pipeline is a substantial operation which is relative to the total size of the pipeline definition. Substantially optimizing this part is unlikely in the near future but this is not really the problem here.

### The problem

A single `CompiledPipeline` is created per pipeline definition:

https://github.com/elastic/logstash/blob/5fd5fb2db2ce45a8da3d52f75ceef1b4caa7cdfb/logstash-core/src/main/java/org/logstash/execution/JavaBasePipelineExt.java#L48

but a `CompileExecution` is actually repeated **per worker** :

https://github.com/elastic/logstash/blob/5fd5fb2db2ce45a8da3d52f75ceef1b4caa7cdfb/logstash-core/src/main/java/org/logstash/config/ir/CompiledPipeline.java#L116-L123

https://github.com/elastic/logstash/blob/5fd5fb2db2ce45a8da3d52f75ceef1b4caa7cdfb/logstash-core/src/main/java/org/logstash/execution/WorkerLoop.java#L46

### Solution

It seems that if we could avoid creating a new `CompiledPipeline` per worker it would solve the problem by having a single actual compilation. Currently separate `CompiledPipeline` are required per worker because they hold per worker state. I have still to better understand that part. 

I tried to see if a trivial shallow cloning of the `CompiledPipeline` per worker would work and while it does significantly reduce the startup time, many java_pipeline specs are failing. (For example a test pipeline on a single worker would take 20s to start and with 128 workers it would take 2m30s but with my shallow clone patch with 128 workers it would take approximately the same time as a single worker, which makes sense). 

I will continue investigating this to see if/how we could avoid this large per worker compilation cost.

**Comment by peacand** [comment]
@colinsurprenant 

Thank you very much for your research, it sounds very good ! In my case it makes sense because the underlying machine is pretty strong and therefore there are 40 workers, so the pipeline is compiled 40x at startup. 

To clarify, the startup time is not the only issue here. For me the issue is that the inputs are actually ready very fast and begin to fetch data while the filters take ages to get ready. In my case, the log sources will detect the TCP ports of Logstash are ready and begin to send data. The question is: what happens in this case where the inputs are flooded with data while the filters are not ready ? I suppose the queues are going to fill up and some logs will finally be lost. No ?

**Comment by colinsurprenant** [comment]
@peacand 
- Right, the workers are started in separate threads and right after the inputs are started. Historically this has never been an issue because there where not really any *extra* delay in starting the workers but now with the Java compilation the inputs can in fact start before the workers have finished their initialization. I have a specific issue for that listed in #11175 and it shouldn't be too hard to fix.
- There are two possibilities here when that occurs: if using persistent queue or not. If you have PQ enabled the inputs will be able to push to the queue and eventually the workers will be able to pickup fro the queue when they are ready.  If not using PQ, the inputs will block trying to push to the memory queue and thus will stop reading. The end result will be dependant on whatever system is connected to that input and its ability to deal with such back pressure.

**Comment by peacand** [comment]
Perfect thank you @colinsurprenant  for your complete answer

Unfortunately in my case, I rely a lot on UDP sources, so there is no way to manage back pressure. And with PQ, I'm afraid that the queues will be so long after ~10 minutes that the delay will never be caught up. 

I look forward the conclusion of your investigation regarding the possibility to compile the pipeline only once for all the workers.

**Comment by colinsurprenant** [comment]
@peacand would waiting for the workers to be fully initialized before starting the inputs be helpful in your situation? Starting the inputs ASAP in conjunction with PQ has the advantage of minimizing data loss at the cost of creating a backlog. OTOH waiting for workers to be ready before starting the inputs avoids potential startup back pressure situations which, depending on your data sources, might be better. We could certainly make that behaviour optional.

**Comment by peacand** [comment]
@colinsurprenant 

I completely agree with your analyze and conclusion !

To be honest I cannot say I will be happy if the inputs wait for the workers to be ready before starting reading events. It would be better, sure, but it also means I will lose > 10 minutes of logs. I will be happy if the overall startup time get back as it was before the Java engine 😊

**Comment by colinsurprenant** [comment]
I just created a PR #11479 to address the redundant compilation per worker. This should significantly reduce the startup time now that the pipeline compilation time is not multiplied by the number of workers. 

If anyone here is able to build and test this PR I would really appreciate your feedback.

**Comment by colinsurprenant** [comment]
I created an alternate implementation #11482 which is probably saner than the one above if anyone wanted to test it.

**Comment by colinsurprenant** [comment]
We just merged https://github.com/elastic/logstash/pull/11482 which should significantly improve java execution pipeline compilation time. This will be included in 7.6.0 and 7.5.2.
This issue is now closed, please feel free to reopen if necessary.

**Comment by peacand** [comment]
Thank you very much for your help and work @colinsurprenant !

I'm looking forward to test in 7.5.2.

**Comment by peacand** [comment]
Wow I've been testing a few deployments I have in production the new version 7.5.2 with your fix. It's perfect, the startup time is back as it was before the Java engine. And the inputs start listening only when the filters are ready, so the pipeline is ready when the inputs starts ingesting events. 

All good and thank you again for your job @colinsurprenant !

**Comment by colinsurprenant** [comment]
Thanks a lot for confirming the fix @peacand! Really glad it improved usability for you.

**Comment by colinsurprenant** [comment]
BTW, about the behaviour of inputs waiting for the workers to be initialized; there is an open issue #11493 to evaluate/discuss adding an option to eagerly start inputs; would love to get feedback on that from users if this is something that might be useful.

**Comment by colinsurprenant** [comment]
Note that the above fix #11482 solved the compilation time problem for a single pipeline use-case but introduced a regression for the multi-pipelines use-case. It has finally been correctly re-fixed in #11564 and is included starting in 7.6 (and in a possible 7.5.3 if that happens).
